File: src/lineage/speciation.py
"""Speciation tracking: species assignment, divergence, extinction events."""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from .genome import Genome
    from .world import Organism


@dataclass
class Species:
    """A species is a group of genetically compatible organisms."""

    id: int
    representative_genome: Genome
    member_ids: set[int] = field(default_factory=set)
    created_at_tick: int = 0
    extinct_at_tick: int | None = None
    parent_species_id: int | None = None
    generation_spawned: int = 0

    @property
    def is_extinct(self) -> bool:
        return self.extinct_at_tick is not None

    @property
    def population(self) -> int:
        return len(self.member_ids)

    def to_dict(self) -> dict:
        return {
            "id": self.id,
            "member_count": len(self.member_ids),
            "created_at_tick": self.created_at_tick,
            "extinct_at_tick": self.extinct_at_tick,
            "parent_species_id": self.parent_species_id,
            "generation_spawned": self.generation_spawned,
        }


@dataclass
class SpeciationEvent:
    """Record of a speciation event."""

    tick: int
    new_species_id: int
    parent_species_id: int | None
    reason: str


@dataclass
class ExtinctionEvent:
    """Record of an extinction event."""

    tick: int
    species_id: int
    final_population: int
    lifespan_ticks: int


class SpeciesTracker:
    """Tracks species assignments, speciation, and extinction."""

    def __init__(self, threshold: float = 0.3) -> None:
        self.threshold = threshold
        self.species: dict[int, Species] = {}
        self.next_species_id = 1
        self.speciation_events: list[SpeciationEvent] = []
        self.extinction_events: list[ExtinctionEvent] = []
        self.phylogeny: dict[int, int | None] = {}

    def assign_species(self, organism: Organism, tick: int) -> int:
        """Assign a species to an organism. Returns the species ID."""
        genome = organism.genome

        best_species_id: int | None = None
        best_distance = float("inf")

        for species_id, species in self.species.items():
            if species.is_extinct:
                continue
            dist = genome.genetic_distance(species.representative_genome)
            if dist < self.threshold and dist < best_distance:
                best_distance = dist
                best_species_id = species_id

        if best_species_id is not None:
            self.species[best_species_id].member_ids.add(organism.id)
            organism.genome.species_id = best_species_id
            return best_species_id

        new_id = self._create_new_species(genome, tick, None)
        self.species[new_id].member_ids.add(organism.id)
        organism.genome.species_id = new_id
        return new_id

    def _create_new_species(
        self, genome: Genome, tick: int, parent_id: int | None
    ) -> int:
        species_id = self.next_species_id
        self.next_species_id += 1

        species = Species(
            id=species_id,
            representative_genome=genome,
            created_at_tick=tick,
            parent_species_id=parent_id,
            generation_spawned=genome.generation,
        )
        self.species[species_id] = species
        self.phylogeny[species_id] = parent_id

        self.speciation_events.append(
            SpeciationEvent(
                tick=tick,
                new_species_id=species_id,
                parent_species_id=parent_id,
                reason="genetic_divergence",
            )
        )
        return species_id

    def update_species(self, organisms: list[Organism], tick: int) -> list[str]:
        """Update species membership, detect extinctions and new splits.

        Returns list of event messages.
        """
        events: list[str] = []
        alive_ids = {o.id for o in organisms if o.alive}

        for species in self.species.values():
            if species.is_extinct:
                continue

            old_members = species.member_ids
            new_members = old_members & alive_ids

            if not new_members:
                species.extinct_at_tick = tick
                lifespan = tick - species.created_at_tick
                self.extinction_events.append(
                    ExtinctionEvent(
                        tick=tick,
                        species_id=species.id,
                        final_population=0,
                        lifespan_ticks=lifespan,
                    )
                )
                events.append(
                    f"Species {species.id} went extinct (t={tick}, "
                    f"lifespan={lifespan})"
                )
            else:
                species.member_ids = new_members

        species_organisms: dict[int, list[Organism]] = {}
        for org in organisms:
            if org.alive:
                sid = org.genome.species_id
                species_organisms.setdefault(sid, []).append(org)

        for species_id, members in species_organisms.items():
            if species_id not in self.species:
                continue
            if self.species[species_id].is_extinct:
                continue

            if len(members) < 2:
                continue

            rep = self.species[species_id].representative_genome
            diverged: list[Organism] = []
            for org in members:
                dist = org.genome.genetic_distance(rep)
                if dist >= self.threshold * 1.5:
                    diverged.append(org)

            if diverged and len(diverged) < len(members):
                new_id = self._create_new_species(
                    diverged[0].genome, tick, species_id
                )
                for org in diverged:
                    self.species[species_id].member_ids.discard(org.id)
                    self.species[new_id].member_ids.add(org.id)
                    org.genome.species_id = new_id

                events.append(
                    f"New species {new_id} diverged from species {species_id} "
                    f"(t={tick}, {len(diverged)} members)"
                )

        return events

    def get_species_count(self) -> int:
        return sum(1 for s in self.species.values() if not s.is_extinct)

    def get_species_populations(self) -> dict[int, int]:
        return {
            s.id: s.population
            for s in self.species.values()
            if not s.is_extinct
        }

    def get_phylogenetic_tree_ascii(self, max_depth: int = 5) -> str:
        """Generate ASCII art phylogenetic tree."""
        lines: list[str] = []
        lines.append("Phylogenetic Tree")
        lines.append("=" * 30)

        roots = [
            sid for sid, parent in self.phylogeny.items() if parent is None
        ]

        def render_node(sid: int, depth: int, prefix: str) -> None:
            if depth > max_depth:
                return
            species = self.species.get(sid)
            if not species:
                return
            status = "EXTINCT" if species.is_extinct else f"pop={species.population}"
            lines.append(f"{prefix}Species {sid} [{status}]")
            children = [
                child for child, parent in self.phylogeny.items() if parent == sid
            ]
            for i, child in enumerate(children):
                is_last = i == len(children) - 1
                child_prefix = prefix + ("    " if is_last else "│   ")
                branch = "└── " if is_last else "├── "
                render_node(child, depth + 1, child_prefix[:-4] + branch)

        for root in roots:
            render_node(root, 0, "")
            lines.append("")

        return "\n".join(lines)

    def to_dict(self) -> dict:
        return {
            "threshold": self.threshold,
            "species": {
                sid: s.to_dict() for sid, s in self.species.items()
            },
            "speciation_events": [
                {
                    "tick": e.tick,
                    "new_species_id": e.new_species_id,
                    "parent_species_id": e.parent_species_id,
                    "reason": e.reason,
                }
                for e in self.speciation_events
            ],
            "extinction_events": [
                {
                    "tick": e.tick,
                    "species_id": e.species_id,
                    "final_population": e.final_population,
                    "lifespan_ticks": e.lifespan_ticks,
                }
                for e in self.extinction_events
            ],
            "phylogeny": self.phylogeny,
        }

File: web/src/simulation/speciation.ts
/**
 * Speciation tracking: species assignment, divergence, extinction events.
 * Port of Python src/lineage/speciation.py — matches behavior exactly.
 */

import {
  type Species,
  type SpeciationEvent,
  type ExtinctionEvent,
  type Organism,
  type Genome,
  Shape,
} from "./types";
import { geneticDistance } from "./genome";

// ── Species helper ──────────────────────────────────────────────────────

/** Check if a species is extinct. */
function isExtinct(species: Species): boolean {
  return species.extinctAtTick !== null;
}

/** Get current population of a species. */
function population(species: Species): number {
  return species.memberIds.size;
}

/** Serialize a Species to a dict matching Python's Species.to_dict format. */
function speciesToDict(species: Species): Record<string, unknown> {
  return {
    id: species.id,
    member_count: species.memberIds.size,
    created_at_tick: species.createdAtTick,
    extinct_at_tick: species.extinctAtTick,
    parent_species_id: species.parentSpeciesId,
    generation_spawned: species.generationSpawned,
  };
}

// ── SpeciesTracker class ────────────────────────────────────────────────

export class SpeciesTracker {
  threshold: number;
  species: Map<number, Species> = new Map();
  nextSpeciesId: number = 1;
  speciationEvents: SpeciationEvent[] = [];
  extinctionEvents: ExtinctionEvent[] = [];
  phylogeny: Map<number, number | null> = new Map();

  constructor(threshold: number = 0.3) {
    this.threshold = threshold;
  }

  /**
   * Assign a species to an organism. Returns [species_id, is_new].
   * Matches Python assign_species exactly.
   */
  assignSpecies(organism: Organism, tick: number): [number, boolean] {
    const genome = organism.genome;

    let bestSpeciesId: number | null = null;
    let bestDistance = Infinity;

    for (const [speciesId, species] of this.species) {
      if (isExtinct(species)) continue;
      const dist = geneticDistance(genome, species.representativeGenome);
      if (dist < this.threshold && dist < bestDistance) {
        bestDistance = dist;
        bestSpeciesId = speciesId;
      }
    }

    if (bestSpeciesId !== null) {
      this.species.get(bestSpeciesId)!.memberIds.add(organism.id);
      organism.genome.speciesId = bestSpeciesId;
      return [bestSpeciesId, false];
    }

    const newId = this.createNewSpecies(genome, tick, null);
    this.species.get(newId)!.memberIds.add(organism.id);
    organism.genome.speciesId = newId;
    return [newId, true];
  }

  /**
   * Create a new species. Returns the new species ID.
   * Matches Python _create_new_species exactly.
   */
  createNewSpecies(
    genome: Genome,
    tick: number,
    parentId: number | null,
  ): number {
    const speciesId = this.nextSpeciesId;
    this.nextSpeciesId += 1;

    const species: Species = {
      id: speciesId,
      representativeGenome: genome,
      memberIds: new Set(),
      createdAtTick: tick,
      extinctAtTick: null,
      parentSpeciesId: parentId,
      generationSpawned: genome.generation,
    };
    this.species.set(speciesId, species);
    this.phylogeny.set(speciesId, parentId);

    this.speciationEvents.push({
      tick,
      newSpeciesId: speciesId,
      parentSpeciesId: parentId,
      reason: "genetic_divergence",
    });

    return speciesId;
  }

  /**
   * Update species membership, detect extinctions and new splits.
   * Returns list of event messages.
   * Matches Python update_species exactly.
   */
  updateSpecies(organisms: Organism[], tick: number): string[] {
    const events: string[] = [];
    const aliveIds = new Set(
      organisms.filter((o) => o.alive).map((o) => o.id),
    );

    // Phase 1 — Extinction detection
    for (const species of this.species.values()) {
      if (isExtinct(species)) continue;

      const oldMembers = species.memberIds;
      const newMembers = new Set<number>();
      for (const id of oldMembers) {
        if (aliveIds.has(id)) {
          newMembers.add(id);
        }
      }

      if (newMembers.size === 0) {
        species.extinctAtTick = tick;
        const lifespan = tick - species.createdAtTick;
        this.extinctionEvents.push({
          tick,
          speciesId: species.id,
          finalPopulation: 0,
          lifespanTicks: lifespan,
        });
        events.push(
          `Species ${species.id} went extinct (t=${tick}, lifespan=${lifespan})`,
        );
      } else {
        species.memberIds = newMembers;
      }
    }

    // Phase 2 — Divergence detection
    const speciesOrganisms: Map<number, Organism[]> = new Map();
    for (const org of organisms) {
      if (org.alive) {
        const sid = org.genome.speciesId;
        let list = speciesOrganisms.get(sid);
        if (!list) {
          list = [];
          speciesOrganisms.set(sid, list);
        }
        list.push(org);
      }
    }

    for (const [speciesId, members] of speciesOrganisms) {
      if (!this.species.has(speciesId)) continue;
      const species = this.species.get(speciesId)!;
      if (isExtinct(species)) continue;

      if (members.length < 2) continue;

      const rep = species.representativeGenome;
      const diverged: Organism[] = [];
      for (const org of members) {
        const dist = geneticDistance(org.genome, rep);
        if (dist >= this.threshold * 1.2) {
          diverged.push(org);
        }
      }

      if (diverged.length > 0 && diverged.length < members.length) {
        const newId = this.createNewSpecies(diverged[0]!.genome, tick, speciesId);
        for (const org of diverged) {
          this.species.get(speciesId)!.memberIds.delete(org.id);
          this.species.get(newId)!.memberIds.add(org.id);
          org.genome.speciesId = newId;
        }

        events.push(
          `New species ${newId} diverged from species ${speciesId} (t=${tick}, ${diverged.length} members)`,
        );
      } else if (members.length > 0 && tick % 100 === 0) {
        species.representativeGenome = members[0]!.genome;
      }
    }

    return events;
  }

  /**
   * Number of non-extinct species. Matches Python get_species_count.
   */
  getSpeciesCount(): number {
    let count = 0;
    for (const species of this.species.values()) {
      if (!isExtinct(species)) count++;
    }
    return count;
  }

  /**
   * Map of species_id → population count for non-extinct species.
   * Matches Python get_species_populations.
   */
  getSpeciesPopulations(): Map<number, number> {
    const result = new Map<number, number>();
    for (const species of this.species.values()) {
      if (!isExtinct(species)) {
        result.set(species.id, population(species));
      }
    }
    return result;
  }

  /**
   * Get the phylogenetic tree structure.
   * Returns map of species_id → { parent, species } for tree building.
   */
  getPhylogeneticTree(): Map<
    number,
    { species: Species; parentSpeciesId: number | null }
  > {
    const result = new Map<
      number,
      { species: Species; parentSpeciesId: number | null }
    >();
    for (const [speciesId, parentId] of this.phylogeny) {
      const species = this.species.get(speciesId);
      if (species) {
        result.set(speciesId, { species, parentSpeciesId: parentId });
      }
    }
    return result;
  }

  /**
   * Generate ASCII art phylogenetic tree. Matches Python get_phylogenetic_tree_ascii exactly.
   */
  getPhylogeneticTreeAscii(maxDepth: number = 5): string {
    const lines: string[] = [];
    lines.push("Phylogenetic Tree");
    lines.push("=".repeat(30));

    // Find roots — species with no parent
    const roots: number[] = [];
    for (const [sid, parent] of this.phylogeny) {
      if (parent === null) {
        roots.push(sid);
      }
    }

    const renderNode = (sid: number, depth: number, prefix: string): void => {
      if (depth > maxDepth) return;
      const species = this.species.get(sid);
      if (!species) return;
      const status = isExtinct(species)
        ? "EXTINCT"
        : `pop=${population(species)}`;
      lines.push(`${prefix}Species ${sid} [${status}]`);

      // Find children
      const children: number[] = [];
      for (const [child, parent] of this.phylogeny) {
        if (parent === sid) {
          children.push(child);
        }
      }

      for (let i = 0; i < children.length; i++) {
        const child = children[i]!;
        const isLast = i === children.length - 1;
        const childPrefix = prefix + (isLast ? "    " : "│   ");
        const branch = isLast ? "└── " : "├── ";
        renderNode(child, depth + 1, childPrefix.slice(0, -4) + branch);
      }
    };

    for (const root of roots) {
      renderNode(root, 0, "");
      lines.push("");
    }

    return lines.join("\n");
  }

  /**
   * Serialize tracker state to dict. Matches Python to_dict exactly.
   */
  toDict(): Record<string, unknown> {
    const speciesDict: Record<string, unknown> = {};
    for (const [sid, s] of this.species) {
      speciesDict[sid] = speciesToDict(s);
    }

    // Convert phylogeny Map to plain object
    const phylogenyObj: Record<string, unknown> = {};
    for (const [sid, parentId] of this.phylogeny) {
      phylogenyObj[sid] = parentId;
    }

    return {
      threshold: this.threshold,
      species: speciesDict,
      speciation_events: this.speciationEvents.map((e) => ({
        tick: e.tick,
        new_species_id: e.newSpeciesId,
        parent_species_id: e.parentSpeciesId,
        reason: e.reason,
      })),
      extinction_events: this.extinctionEvents.map((e) => ({
        tick: e.tick,
        species_id: e.speciesId,
        final_population: e.finalPopulation,
        lifespan_ticks: e.lifespanTicks,
      })),
      phylogeny: phylogenyObj,
    };
  }

  /**
   * Deserialize tracker state from dict. Matches Python from_dict pattern.
   * Note: Python doesn't have a from_dict on SpeciesTracker, but we need
   * it for loading recordings.
   */
  static fromDict(d: Record<string, unknown>): SpeciesTracker {
    const tracker = new SpeciesTracker(d["threshold"] as number);

    // Restore species
    const speciesData = d["species"] as Record<string, Record<string, unknown>>;
    for (const [sidStr, sData] of Object.entries(speciesData)) {
      const sid = Number(sidStr);
      // We need a placeholder genome since to_dict doesn't store it
      // The full genome would come from organism data
      const placeholderGenome: Genome = {
        size: 0,
        shape: Shape.CIRCLE,
        colorHue: 0,
        colorSat: 0,
        colorVal: 0,
        baseMetabolism: 0,
        photosynthesisRate: 0,
        movementSpeed: 0,
        foodDetectionRadius: 0,
        predatorDetectionRadius: 0,
        mateDetectionRadius: 0,
        reproductionThreshold: 0,
        reproductionCost: 0,
        sexualReproduction: false,
        crossoverPreference: 0,
        mutationRate: 0,
        mutationMagnitude: 0,
        behaviorTree: [],
        speciesId: sid,
        generation: (sData["generation_spawned"] as number) ?? 0,
        parentIds: [],
      };
      tracker.species.set(sid, {
        id: sid,
        representativeGenome: placeholderGenome,
        memberIds: new Set<number>(), // member_ids not stored in to_dict
        createdAtTick: (sData["created_at_tick"] as number) ?? 0,
        extinctAtTick: (sData["extinct_at_tick"] as number | null) ?? null,
        parentSpeciesId:
          (sData["parent_species_id"] as number | null) ?? null,
        generationSpawned: (sData["generation_spawned"] as number) ?? 0,
      });
      tracker.nextSpeciesId = Math.max(tracker.nextSpeciesId, sid + 1);
    }

    // Restore speciation events
    const specEvents = d["speciation_events"] as Record<string, unknown>[];
    if (specEvents) {
      tracker.speciationEvents = specEvents.map((e) => ({
        tick: e["tick"] as number,
        newSpeciesId: e["new_species_id"] as number,
        parentSpeciesId: e["parent_species_id"] as number | null,
        reason: e["reason"] as string,
      }));
    }

    // Restore extinction events
    const extEvents = d["extinction_events"] as Record<string, unknown>[];
    if (extEvents) {
      tracker.extinctionEvents = extEvents.map((e) => ({
        tick: e["tick"] as number,
        speciesId: e["species_id"] as number,
        finalPopulation: e["final_population"] as number,
        lifespanTicks: e["lifespan_ticks"] as number,
      }));
    }

    // Restore phylogeny
    const phylogenyData = d["phylogeny"] as Record<string, unknown>;
    if (phylogenyData) {
      for (const [sidStr, parentId] of Object.entries(phylogenyData)) {
        tracker.phylogeny.set(Number(sidStr), parentId as number | null);
      }
    }

    return tracker;
  }
}
